oauth2_http/javatests/com/google/auth/oauth2/GdchCredentialsTest.java

```diff
@@ -0,0 +1,1002 @@
+    GenericJson json =
+        writeGdchServiceAccountJson(
+            FORMAT_VERSION,
+            PROJECT_ID,
+            PRIVATE_KEY_ID,
+            PRIVATE_KEY_PKCS8,
+            SERVICE_IDENTITY_NAME,
+            "/path/to/missing/file",
+            TOKEN_SERVER_URI);
+
+    try {
+      GdchCredentials credentials = GdchCredentials.fromJson(json, transportFactory);
+      fail("Should not be able to create GDCH credential without exception.");
+    } catch (IOException ex) {
+      assertTrue(ex.getMessage().contains("Error reading certificate file from CA cert path"));
+    }
+  }
+
+  @Test
+  public void fromJSON_nullTokenServerUri() throws IOException {
+    MockTokenServerTransportFactory transportFactory = new MockTokenServerTransportFactory();
+    GenericJson json =
+        writeGdchServiceAccountJson(
+            FORMAT_VERSION,
+            PROJECT_ID,
+            PRIVATE_KEY_ID,
+            PRIVATE_KEY_PKCS8,
+            SERVICE_IDENTITY_NAME,
+            CA_CERT_PATH,
+            null);
+
+    try {
+      GdchCredentials credentials = GdchCredentials.fromJson(json, transportFactory);
+      fail("Should not be able to create GDCH credential without exception.");
+    } catch (IOException ex) {
+      assertTrue(
+          ex.getMessage()
+              .contains(
+                  String.format(
+                      "Error reading GDCH service account credential from JSON, "
+                          + "%s is misconfigured.",
+                      "token_uri")));
+    }
+  }
+
+  @Test
+  public void fromJSON_defaultTransportFactory() throws IOException {
+    MockTokenServerTransportFactory transportFactory = new MockTokenServerTransportFactory();
+    GenericJson json =
+        writeGdchServiceAccountJson(
+            FORMAT_VERSION,
+            PROJECT_ID,
+            PRIVATE_KEY_ID,
+            PRIVATE_KEY_PKCS8,
+            SERVICE_IDENTITY_NAME,
+            null,
+            TOKEN_SERVER_URI);
+    GdchCredentials credentials = GdchCredentials.fromJson(json, transportFactory);
+    assertEquals(
+        MockTokenServerTransportFactory.class, credentials.getTransportFactory().getClass());
+  }
+
+  @Test
+  public void fromJSON_transportFactoryForGdch() throws IOException {
+    MockTokenServerTransportFactory transportFactory = new MockTokenServerTransportFactory();
+    GenericJson json =
+        writeGdchServiceAccountJson(
+            FORMAT_VERSION,
+            PROJECT_ID,
+            PRIVATE_KEY_ID,
+            PRIVATE_KEY_PKCS8,
+            SERVICE_IDENTITY_NAME,
+            CA_CERT_PATH,
+            TOKEN_SERVER_URI);
+    GdchCredentials credentials = GdchCredentials.fromJson(json, transportFactory);
+    assertEquals(
+        GdchCredentials.TransportFactoryForGdch.class,
+        credentials.getTransportFactory().getClass());
+  }
+
+  @Test
+  public void fromJSON_hasAccessToken() throws IOException {
```

Comment:
In this case, I think we should use the transportFactory with trust store (i.e. TransportFactoryForGdch) to get request metadata, but the logic here is using the default transportFactory. Should I create a new mock token server for this? any suggestions?

Reply:
I don't think this is necessary, you already have tests to cover the usage of gdch transport factory in from JSON.